docs/building.md

```diff
@@ -0,0 +1,228 @@
+# Proxy-Wasm C++ SDK Build Instructions
+
+The C++ SDK has dependencies on specific versions of the C++ WebAssembly
+toolchain [Emscripten](https://emscripten.org) and the protobuf library,
+therefore use of a Docker image is recommended.
+
+## Docker
+
+A Dockerfile for the C++ SDK is provided in [Dockerfile-sdk](Dockerfile-sdk).
+
+It can built in this directory by:
+
+```bash
+docker build -t wasmsdk:v2 -f Dockerfile-sdk .
+```
+
+The docker image can be used for compiling C++ plugin code into Wasm modules.
+
+### Creating a project for use with the Docker build image
+
+Create a directory with your source files and a Makefile:
+
+```makefile
+PROXY_WASM_CPP_SDK=/sdk
+
+all: myproject.wasm
+
+include ${PROXY_WASM_CPP_SDK}/Makefile.base_lite
+```
+
+Source file (myproject.cc):
+
+```c++
+#include <string>
+#include <unordered_map>
+
+#include "proxy_wasm_intrinsics.h"
+
+class ExampleContext : public Context {
+public:
+  explicit ExampleContext(uint32_t id, RootContext* root) : Context(id, root) {}
+
+  FilterHeadersStatus onRequestHeaders(uint32_t headers, bool end_of_stream) override;
+  void onDone() override;
+};
+static RegisterContextFactory register_ExampleContext(CONTEXT_FACTORY(ExampleContext));
+
+FilterHeadersStatus ExampleContext::onRequestHeaders(uint32_t headers, bool end_of_stream) {
+  logInfo(std::string("onRequestHeaders ") + std::to_string(id()));
+  auto path = getRequestHeader(":path");
+  logInfo(std::string("header path ") + std::string(path->view()));
+  return FilterHeadersStatus::Continue;
+}
+
+void ExampleContext::onDone() { logInfo("onDone " + std::to_string(id())); }
+```
+
+### Compiling with the Docker build image
+
+Run docker:
+
+```bash
+docker run -v $PWD:/work -w /work  wasmsdk:v2 /build_wasm.sh
+```
+
+### Caching the standard libraries
+
+The first time that emscripten runs it will generate the standard libraries.  To
+cache these in the docker image, after the first successful compilation (e.g
+myproject.cc above), commit the image with the standard libraries:
+
+```bash
+docker commit `docker ps -l | grep wasmsdk:v2 | awk '{print $1}'` wasmsdk:v2
+```
+
+This will save time on subsequent compiles.
+
+### Using the SDK from a newer/specific version of Envoy
+
+To use a newer/specific version of the SDK (e.g. from the version of Envoy you
+are going to deploy the WebAssembly module to), bind that volume and use it in
+the Makefile.
+
+Here is an example Makefile referencing the SDK at `../envoy/api/wasm/cpp` and
+mounted as `sdk` in the `/work` directory:
+
+```makefile
+PROXY_WASM_CPP_SDK=/work/sdk
+
+all: myproject.wasm
+
+include ${PROXY_WASM_CPP_SDK}/Makefile.base_lite
+```
+
+Run docker pointing to Envoy sources in a directory parallel (at the same level)
+as your project directory:
+
+```bash
+docker run -v $PWD:/work -v $PWD/../envoy/api/wasm/cpp:/work/sdk -w /work  wasmsdk:v2 bash /build_wasm.sh
+```
+
+### Using Abseil from the Docker image
```

Comment:
I don't think we're using Abseil in C++ SDK anymore.

Reply:
Currently, sdk_container.sh still bundles a version of Abseil into the Docker image. Rather than change any behavior in this PR, I'd prefer to keep mention of Abseil until we remove that.